futures-util/src/future/join_all.rs

```diff
@@ -19,13 +20,30 @@ fn iter_pin_mut<T>(slice: Pin<&mut [T]>) -> impl Iterator<Item = Pin<&mut T>> {
     unsafe { slice.get_unchecked_mut() }.iter_mut().map(|t| unsafe { Pin::new_unchecked(t) })
 }
 
-/// Future for the [`join_all`] function.
 #[must_use = "futures do nothing unless you `.await` or poll them"]
-pub struct JoinAll<F>
-where
-    F: Future,
-{
-    elems: Pin<Box<[MaybeDone<F>]>>,
+pin_project_lite::pin_project! {
+    /// Future for the [`join_all`] function.
+    pub struct JoinAll<F>
+    where
+        F: Future,
+    {
+        #[pin]
+        kind: JoinAllKind<F>,
+    }
+}
+
+const SMALL: usize = 30;
+
+pin_project_lite::pin_project! {
+    #[project = JoinAllKindProj]
+    pub enum JoinAllKind<F>
+    where
+        F: Future,
+    {
+        Small { elems: Pin<Box<[MaybeDone<F>]>> },
+        #[cfg(not(futures_no_atomic_cas))]
+        Big  { #[pin] fut: Collect<FuturesOrdered<F>, Vec<F::Output>> },
+    }
 }
```

Comment:
IIUC, both `Pin<Box<[MaybeDone<F>]>>` and `Collect<FuturesOrdered<F>` are always `Unpin`, so `JoinAllKind` is always `Unpin` and `pin_project!` is not needed here.